### PR TITLE
Set PeerReady condition status to false upon relocation initiation

### PR DESCRIPTION
**Pull request by BenamarMk**
Previously, the `PeerReady` condition status was set to `false` after the relocation setup between the source and target clusters was completed and before entering the clean-up phase. This led to a misleading indication that the peer was ready for a new action before getting to that point. Consequently, users can initiate another action, assuming that the `PeerReady` condition being `true` meant it was ready for the next step.

To rectify this issue, this commit ensures that as soon as the relocation action is initiated, the `PeerReady` condition status is immediately switched to `false`. This change accurately reflects the actual readiness of the peer, preventing any misconceptions about initiating additional actions prematurely.

Fixes: bz [2138855](https://bugzilla.redhat.com/show_bug.cgi?id=2138855)

### CODE REVIEW

File: controllers/drplacementcontrol.go [hunk 1]
```diff
@@ -741,7 +741,7 @@ func (d *DRPCInstance) RunRelocate() (bool, error) {
 		return !done, fmt.Errorf(errMsg)
 	}
 
-	if !d.validatePeerReady() {
+	if d.getLastDRState() != rmn.Relocating && !d.validatePeerReady() {
```
Comment (nirs) [inline controllers/drplacementcontrol.go:744]: Is this required because we set peer ready to false earlier, or additional change? If another change, can you separate it to another commit and explain the reason for this change?

File: controllers/drplacementcontrol.go [hunk 2]
```diff
@@ -1048,6 +1048,9 @@ func (d *DRPCInstance) relocate(preferredCluster, preferredClusterNamespace stri
 	d.setDRState(drState)
```
Comment (nirs) [inline controllers/drplacementcontrol.go:1048]: The comment above:

    // Make sure we record the state that we are failing over

Is copy paste from failover code?

File: controllers/drplacementcontrol.go [hunk 3]
```diff
@@ -1048,6 +1048,9 @@ func (d *DRPCInstance) relocate(preferredCluster, preferredClusterNamespace stri
...
 	addOrUpdateCondition(&d.instance.Status.Conditions, rmn.ConditionAvailable, d.instance.Generation,
 		d.getConditionStatusForTypeAvailable(), string(d.instance.Status.Phase), "Starting relocation")
+	addOrUpdateCondition(&d.instance.Status.Conditions, rmn.ConditionPeerReady, d.instance.Generation,
+		metav1.ConditionFalse, rmn.ReasonNotStarted,
+		fmt.Sprintf("Started relocation to cluster %q", preferredCluster))
```
Comment (nirs) [inline controllers/drplacementcontrol.go:1053]: The reason "ReasonNotStarted" contradicts the test "Started relocation".

**Review by nirs — COMMENTED** [review]
Setting peer ready early looks good to me, but I don't understand the second change.

Also the title of the commit contradicts the commit message and the code - do you mean "Set PeerReady condition status to *false* upon relocation initiation"?

**Review by ShyamsundarR — COMMENTED** [review]
LGTM

An orthogonal question/observation:

EnsureCleanup marks PeerReady post its operations are successful. ensureCleanupAndVolSyncReplicationSetup sets up EnsureVolSyncReplicationSetup post cleanup. Should we not mark PeerReady post EnsureVolSyncReplicationSetup is also a success from a VolSync POV?

File: controllers/drplacementcontrol.go (hunk 1 above)
Comment (ShyamsundarR) [reply controllers/drplacementcontrol.go:744]: Reconcile will loop over `Relocate` multiple times as various states change (e.g current primary transitions to secondary etc.). As the PeerReady is set to false for the current generation of the spec, once relocation has started, a future loop will error out here. This additional check hence is required to pass this gate if we checked everything was fine with the prior generation and started relocation.

File: controllers/drplacementcontrol.go (hunk 3 above)
Comment (ShyamsundarR) [reply controllers/drplacementcontrol.go:1053]: We could potentially reword this to "Relocation in progress to cluster %q"

**Comment by BenamarMk** [comment]
> EnsureCleanup marks PeerReady post its operations are successful. ensureCleanupAndVolSyncReplicationSetup sets up EnsureVolSyncReplicationSetup post cleanup. Should we not mark PeerReady post EnsureVolSyncReplicationSetup is also a success from a VolSync POV?

Not sure I understand the question. EnsureCleanup() is where we set the PeerReady as a success. We don't enter the volsync setup without ensuring that the peer is ready.

**Comment by ShyamsundarR** [comment]
> > EnsureCleanup marks PeerReady post its operations are successful. ensureCleanupAndVolSyncReplicationSetup sets up EnsureVolSyncReplicationSetup post cleanup. Should we not mark PeerReady post EnsureVolSyncReplicationSetup is also a success from a VolSync POV?
> 
> Not sure I understand the question. EnsureCleanup() is where we set the PeerReady as a success. We don't enter the volsync setup without ensuring that the peer is ready.

(just reiterating, does not impact this PR)

So if volsync is not setup on the secondary, is it ready for a relocate? IOW should we declare DRPC PeerReady with Volsync protected volumes before the volsync setup is complete?

(above is based on my understanding that we declare PeerReady true in DRPC post EnsureCleanup and pre EnsureVolSyncReplicationSetup)

**Review by ShyamsundarR — APPROVED** [review]

**Review by nirs — APPROVED** [review]
